model/modeldecoder/v2/decoder_test.go

```diff
@@ -0,0 +1,169 @@
+	})
+
+	t.Run("decode", func(t *testing.T) {
+		var out model.Metadata
+		decodeProfileMetadata(&out)
+		assert.Equal(t, model.Metadata{Service: model.Service{
+			Name:  "user-service",
+			Agent: model.Agent{Name: "go", Version: "1.0.0"}}}, out)
+
+		err := DecodeProfileMetadata(decoder.NewJSONIteratorDecoder(strings.NewReader(`malformed`)), &out)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "decode")
+	})
+
+	t.Run("validate", func(t *testing.T) {
+		inp := `{}`
+		var out model.Metadata
+		err := DecodeProfileMetadata(decoder.NewJSONIteratorDecoder(strings.NewReader(inp)), &out)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "validation")
+	})
+
+}
+
+func TestDecodeMetadata(t *testing.T) {
+	var testMinValidMetadata = `
+	{"metadata":{"service":{"name":"user-service","agent":{"name":"go","version":"1.0.0"}}}}`
+
+	decodeMetadata := func(out *model.Metadata) {
+		dec := decoder.NewJSONIteratorDecoder(strings.NewReader(testMinValidMetadata))
+		require.NoError(t, DecodeMetadata(dec, out))
+	}
+
+	t.Run("fetch-release", func(t *testing.T) {
+		// this test cannot be run in parallel with other tests using the
+		// metadataNoKeyPool sync.Pool
+
+		// whenever DecodeProfileMetadata is called a metadata instance
+		// should be retrieved from the sync.Pool and be released on finish
+		// test behavior by overriding the new method, occupying an existing
+		// instance and counting how often the New method is called
+		var newCount, expectedNewCount int
+		origNew := metadataWithKeyPool.New
+		defer func() { metadataWithKeyPool.New = origNew }()
+		metadataWithKeyPool.New = func() interface{} {
+			newCount++
+			return &metadataWithKey{}
+		}
+		var out model.Metadata
+		// on the first call align the expected with the current new count
+		// important since other tests might have run already
+		decodeMetadata(&out)
+		expectedNewCount = newCount
+		// on the second call it should reuse the metadata instance
+		decodeMetadata(&out)
+		assert.Equal(t, expectedNewCount, newCount)
+		// force a new instance on the next decoder call
+		fetchMetadataWithKey()
+		decodeMetadata(&out)
+		assert.Equal(t, expectedNewCount+1, newCount)
+	})
+
+	t.Run("decode", func(t *testing.T) {
+		var out model.Metadata
+		decodeMetadata(&out)
+		assert.Equal(t, model.Metadata{Service: model.Service{
+			Name:  "user-service",
+			Agent: model.Agent{Name: "go", Version: "1.0.0"}}}, out)
+
+		err := DecodeMetadata(decoder.NewJSONIteratorDecoder(strings.NewReader(`malformed`)), &out)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "decode")
+	})
+
+	t.Run("validate", func(t *testing.T) {
+		inp := `{}`
+		var out model.Metadata
+		err := DecodeMetadata(decoder.NewJSONIteratorDecoder(strings.NewReader(inp)), &out)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "validation")
+	})
+
+}
+
+func TestMappingToModel(t *testing.T) {
+	//TODO(simitt)
+	//  override existing values if they are set
```

Comment:
just checking - will you be doing this in a followup, or did you miss this?

Reply:
I forgot about them, added them now and also tests for the `rumv3` package.

Reply:
did you forget to push?

Reply:
pushed the changes now, together with some more test refactoring